sync/class.jetpack-sync-actions.php

```diff
@@ -178,25 +182,63 @@ static function send_data( $data, $codec_name, $sent_timestamp, $queue_id ) {
 	}
 
 	static function do_initial_sync( $new_version = null, $old_version = null ) {
-		$initial_sync_config = array(
+		$initial_sync_config = self::get_update_full_sync_config();
+		$include_users = false;
+		if ( $old_version && ( version_compare( $old_version, '4.2', '<' ) ) ) {
+			$initial_sync_config['users'] = 'initial';
+			$include_users = true;
+		}
+
+		if ( is_multisite() &&
+		     // Is Jetpack network activated
+		     in_array( 'jetpack/jetpack.php' , array_keys( get_site_option( 'active_sitewide_plugins' ) ) ) ) {
+			// stagger initial syncs for multisite blogs so they don't all pile on top of each other
+			if ( is_main_site() ) {
+				wp_schedule_single_event( time() , 'jetpack_full_sync_on_multisite_jetpack_upgrade_cron', array( $include_users ) );
+			}
+		} else {
+			self::do_full_sync( $initial_sync_config );
+		}
+	}
+
+	static function get_update_full_sync_config() {
+		return array(
 			'options' => true,
 			'network_options' => true,
 			'functions' => true,
 			'constants' => true,
 		);
-
-		if ( $old_version && ( version_compare( $old_version, '4.2', '<' ) ) ) {
+	}
+	
+	static function full_sync_on_multisite_jetpack_upgrade( $include_users = false ) {
+		global $wpdb;
+		$initial_sync_config = self::get_update_full_sync_config();
+		if ( $include_users ) {
 			$initial_sync_config['users'] = 'initial';
 		}
 
-		self::do_full_sync( $initial_sync_config );
+		$max_blog_id = 0;
+		$batch_size = 500;
+
+		while( true ) {
```

Comment:
What do we do if the request times out while these syncs are being kicked off? Our current full sync implementation is very resistant to failure, as it picks up where it left off if the process is killed. This method doesn't do that. For large multisites, I can imagine it would be easy for this to run out of memory/time/cpu-cycles and get killed off.

The reason this isn't a problem for the current implementation is that the action `updating_jetpack_version` runs based on state stored in each site. Without storing state somewhere, it's hard to know if we still need to run initial sync for some sites.